File: 115-distinct-subsequences/115-distinct-subsequences.java
class Solution {
    public int numDistinct(String s, String t) {
        int n1 = s.length(), n2 = t.length();
        int[] dp = new int[n2 + 1];
        dp[0] = 1;
        for(int idx1 = 1; idx1 <= n1; ++idx1) {
            int[] current = new int[n2 + 1];
            current[0] = 1;
            for(int idx2 = 1; idx2 <= n2; ++idx2) {
                int take = dp[idx2 - 1];
                int notTake = dp[idx2];
                if(s.charAt(idx1 - 1) == t.charAt(idx2 - 1))
                    current[idx2] = take + notTake;
                else
                    current[idx2] = notTake;
            }
            dp = current;
        }
        return dp[n2];
    }
}

File: 115-distinct-subsequences/115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n1 = len(s)
        n2 = len(t)
        dp = [0] * (n2 + 1)
        dp[0] = 1
        for idx1 in range (1, n1 + 1):
            current = [0] * (n2 + 1)
            current[0] = 1
            for idx2 in range (1, n2 + 1):
                take = dp[idx2 - 1]
                notTake = dp[idx2]
                if s[idx1 - 1] == t[idx2 - 1]:
                    current[idx2] = take + notTake
                else:
                    current[idx2] = notTake
            dp = current
        return dp[n2]